File: supabase/functions/_shared/meal_catalog.ts
// Public catalog filter shared by ai-search / ILIKE fallbacks.
// Only `Available` inventory rows are searchable.

export function isAvailableStatus(raw: unknown): boolean {
  return String(raw ?? "").trim().toLowerCase() === "available"
}

export function isSellableMeal(row: Record<string, unknown> | null | undefined): boolean {
  if (!row) return false
  if (!isAvailableStatus(row.status)) return false
  const customerName = String(row.customer_name ?? "").trim()
  return customerName.length === 0
}

export function filterSellableMeals<T extends Record<string, unknown>>(rows: T[] | null | undefined): T[] {
  if (!rows) return []
  return rows.filter((row) => isSellableMeal(row))
}

File: lib/utils/meal_catalog.dart
// Catalog rows that diners may browse / search.
// Live meals.status is Title Case (`Available`). Archived, Closed, Paused,
// cancelled, and sold-out plates stay out of the public feed and AI search.

abstract final class MealCatalog {
  static const availableStatus = 'Available';

  static const excludedStatuses = <String>{
    'archived',
    'closed',
    'paused',
    'cancelled',
    'canceled',
    'sold out',
    'sold_out',
    'soldout',
  };

  static String _normalize(String? raw) =>
      (raw ?? '').trim().toLowerCase().replaceAll(RegExp(r'[\s_]+'), ' ');

  static bool isAvailableStatus(String? raw) => _normalize(raw) == 'available';

  /// Inventory catalog row (not a legacy "order stored as a meal").
  static bool isSellable(Map<String, dynamic> meal) {
    if (!isAvailableStatus(meal['status']?.toString())) return false;
    final customerName = meal['customer_name']?.toString().trim() ?? '';
    return customerName.isEmpty;
  }
}
